### PR TITLE
[Merged by Bors] - feat: port Analysis.Normed.Group.AddCircle

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Normed/Group/AddCircle.lean [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+  · rcases eq_or_ne p 0 with (rfl | hp)
+    · simp only [abs_zero, zero_div] at hε
+      simp only [not_lt.mpr hε, coe_real_preimage_closedBall_period_zero, abs_zero, zero_div,
+        if_false, inter_eq_right_iff_subset]
+      exact hs.trans (closedBall_subset_closedBall <| by simp [hε])
+    -- Porting note: was
+    -- simp [closedBall_eq_univ_of_half_period_le p hp (↑x) hε, not_lt.mpr hε]
+    simp only [not_lt.mpr hε, ite_false, inter_eq_right_iff_subset]
+    rw [closedBall_eq_univ_of_half_period_le p hp (↑x : ℝ ⧸ zmultiples p) hε]
+    simp
```
Comment (Parcly-Taxel) [inline Mathlib/Analysis/Normed/Group/AddCircle.lean:219]:
```suggestion
    rw [closedBall_eq_univ_of_half_period_le p hp (↑x : ℝ ⧸ zmultiples p) hε, preimage_univ]
    apply subset_univ
```

File: Mathlib/Analysis/Normed/Group/AddCircle.lean [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    rcases hp.symm.lt_or_lt with (hp | hp)
+    · rw [abs_eq_self.mpr hp.le] at hx
+      exact this p hp hx
+    · rw [← norm_neg_period]
+      rw [abs_eq_neg_self.mpr hp.le] at hx
+      exact this (-p) (neg_pos.mpr hp) hx
+  clear hx
+  intro p hp hx
+  -- Porting note: added type annotation
+  rcases eq_or_ne x (p / 2) with ((rfl : x = p / 2) | hx')
```
Comment (Parcly-Taxel) [inline Mathlib/Analysis/Normed/Group/AddCircle.lean:162]:
```suggestion
  rcases eq_or_ne x (p / (2 : ℝ)) with (rfl | hx')
```

File: Mathlib/Analysis/Normed/Group/AddCircle.lean [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+  obtain ⟨x⟩ := x
+  change ‖(x : AddCircle p)‖ ≤ |p| / 2
+  rw [norm_eq, ← mul_le_mul_left (abs_pos.mpr (inv_ne_zero hp)), ← abs_mul, mul_sub, mul_left_comm,
+    ← mul_div_assoc, ← abs_mul, inv_mul_cancel hp, mul_one, abs_one]
+  exact abs_sub_round (p⁻¹ * x)
+#align add_circle.norm_le_half_period AddCircle.norm_le_half_period
+
+@[simp]
+theorem norm_half_period_eq : ‖(↑(p / 2) : AddCircle p)‖ = |p| / 2 := by
+  rcases eq_or_ne p 0 with (rfl | hp); · simp
```
Comment (Parcly-Taxel) [inline Mathlib/Analysis/Normed/Group/AddCircle.lean:144]:
```suggestion
  rcases eq_or_ne p 0 with (rfl | hp); simp
```

File: Mathlib/Analysis/Normed/Group/AddCircle.lean [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  change ∃ k : ℕ, ‖u‖ = p * (k / n)
+  obtain ⟨m, -, -, hm⟩ := exists_gcd_eq_one_of_isOfFinAddOrder hu
+  refine' ⟨min (m % n) (n - m % n), _⟩
+  rw [← hm, norm_div_nat_cast]
+#align add_circle.exists_norm_eq_of_fin_add_order AddCircle.exists_norm_eq_of_isOfFinAddOrder
+
+theorem le_add_order_smul_norm_of_isOfFinAddOrder {u : AddCircle p} (hu : IsOfFinAddOrder u)
+    (hu' : u ≠ 0) : p ≤ addOrderOf u • ‖u‖ := by
+  obtain ⟨n, hn⟩ := exists_norm_eq_of_isOfFinAddOrder hu
+  replace hu : (addOrderOf u : ℝ) ≠ 0;
```
Comment (Parcly-Taxel) [inline Mathlib/Analysis/Normed/Group/AddCircle.lean:271]:
```suggestion
  replace hu : (addOrderOf u : ℝ) ≠ 0
```

File: Mathlib/Analysis/Normed/Group/AddCircle.lean [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+  AddSubgroup.normedAddCommGroupQuotient _
+
+@[simp]
+theorem norm_coe_mul (x : ℝ) (t : ℝ) :
+    ‖(↑(t * x) : AddCircle (t * p))‖ = |t| * ‖(x : AddCircle p)‖ := by
+  have aux : ∀ {a b c : ℝ}, a ∈ zmultiples b → c * a ∈ zmultiples (c * b) := fun {a b c} h => by
+    simp only [mem_zmultiples_iff] at h⊢
+    obtain ⟨n, rfl⟩ := h
+    exact ⟨n, (mul_smul_comm n c b).symm⟩
+  rcases eq_or_ne t 0 with (rfl | ht); · simp
```
[the remark this answers is not in the document]
Comment (Ruben-VandeVelde) [reply Mathlib/Analysis/Normed/Group/AddCircle.lean:53]: If you want to change this, move the `· simp` to the next line and remove the semicolon. You suggestion would make it harder to figure out what's going on if `simp` ever stops proving this goal.

Feel free to push your other suggestions

**Pull request closed without merging**